metalcloud/resource_metalcloud_extension_instance.go

```diff
@@ -0,0 +1,586 @@
+package metalcloud
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/antihax/optional"
+	"github.com/hashicorp/go-cty/cty"
+	log "github.com/hashicorp/terraform-plugin-log/tflog"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	sdk2 "github.com/metalsoft-io/metal-cloud-sdk2-go"
+)
+
+const (
+	_INFRASTRUCTURE_ID = "infrastructure_id"
+
+	_EXTENSION_INSTANCE_ID     = "extension_instance_id"
+	_EXTENSION_INSTANCE_LABEL  = "extension_instance_label"
+	_EXTENSION_INSTANCE_INPUT  = "extension_instance_input"
+	_EXTENSION_INSTANCE_OUTPUT = "extension_instance_output"
+)
+
+func resourceExtensionInstance() *schema.Resource {
+	return &schema.Resource{
+		SchemaFunc:    resourceExtensionInstanceSchema,
+		CreateContext: resourceExtensionInstanceCreate,
+		ReadContext:   resourceExtensionInstanceRead,
+		UpdateContext: resourceExtensionInstanceUpdate,
+		DeleteContext: resourceExtensionInstanceDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+	}
+}
+
+func resourceExtensionInstanceSchema() map[string]*schema.Schema {
+	schema := map[string]*schema.Schema{
+		_EXTENSION_INSTANCE_ID: {
+			Type:     schema.TypeInt,
+			Computed: true,
+		},
+		_INFRASTRUCTURE_ID: {
+			Type:     schema.TypeInt,
+			Required: true,
+			ForceNew: true,
+			ValidateFunc: func(val interface{}, key string) (warns []string, errs []error) {
+				v := val.(int)
+				if v == 0 {
+					errs = append(errs, fmt.Errorf("%q is required. Provided value: %d", key, v))
+				}
+				return
+			},
+		},
+		_EXTENSION_ID: {
+			Type:     schema.TypeInt,
+			Required: true,
+			ForceNew: true,
+			ValidateFunc: func(val interface{}, key string) (warns []string, errs []error) {
+				v := val.(int)
+				if v == 0 {
+					errs = append(errs, fmt.Errorf("%q is required. Provided value: %d", key, v))
+				}
+				return
+			},
+		},
+		_EXTENSION_INSTANCE_LABEL: {
+			Type:     schema.TypeString,
+			Optional: true,
+			Default:  nil,
+			Computed: true,
+			ForceNew: true,
+			//this is required because on the serverside the labels are converted to lowercase automatically
+			DiffSuppressFunc: func(_, old, new string, d *schema.ResourceData) bool {
+				if strings.EqualFold(old, new) {
```

Comment:
Use the utils function https://github.com/metalsoft-io/terraform-provider-metalcloud/blob/master/metalcloud/utils.go#L56